qcfractal/tests/test_collections.py

```diff
@@ -132,6 +212,114 @@ def test_gradient_dataset_statistics(gradient_dataset_fixture):
 
+@pytest.fixture(scope="module")
+def contributed_dataset_fixture(fractal_compute_server):
+    """ Fixture for testing rich contributed datasets with many properties and molecules of different sizes"""
+    client = ptl.FractalClient(fractal_compute_server)
+
+    testing.check_has_module("psi4")
+
+    # Build a dataset
+    ds = ptl.collections.Dataset("ds_contributed",
+                                 client)
+
+    ds.add_entry("He1", ptl.Molecule.from_data("He -1 0 0\n--\nHe 0 0 1"))
+    ds.add_entry("He", ptl.Molecule.from_data("He -1.1 0 0"))
+    ds.units = "hartree"
+    ds.save()
+
+    energy = {
+        "name": "Fake Energy",
+        "theory_level": "pseudo-random values",
+        "values": {
+            "He1": 1234.5,
+            "He": 5.4321
+        },
+        "theory_level_details": {
+            "driver": "energy",
+            "program": "fake_program",
+            "basis": "fake_basis",
+            "method": "fake_method",
+        },
+        "units": "hartree"
+    }
+    gradient = {
+        "name": "Fake Gradient",
+        "theory_level": "pseudo-random values",
+        "values": {
+            "He1": [0.03, 0, 0.02, -0.02, 0, -0.03],
+            "He": [0.03, 0, 0.02]
+        },
+        "theory_level_details": {
+            "driver": "gradient",
+            "program": "fake_program",
+            "basis": "fake_basis",
+            "method": "fake_method",
+        },
+        "units": "hartree/bohr"
+    }
+    hessian = {
+        "name": "Fake Hessian",
+        "theory_level": "pseudo-random values",
+        "values": {
+            "He1": list(np.eye(6).ravel()),
+            "He": [1, 0.2, 0.1, 0.2, 1, 0.4, 0.1, 0.4, 1]
+        },
+        "theory_level_details": {
+            "driver": "hessian",
+            "program": "fake_program",
+            "basis": "fake_basis",
+            "method": "fake_method",
+        },
+        "units": "hartree/bohr**2"
+    }
+    dipole = {
+        "name": "Fake Dipole",
+        "theory_level": "pseudo-random values",
+        "values": {
+            "He": [1., 2., 3.],
+            "He1": [1., -2., 0.]
+        },
+        "theory_level_details": {
+            "driver": "dipole",
+            "program": "fake_program",
+            "basis": "fake_basis",
+            "method": "fake_method",
+        },
+        "units": "e * bohr"
+    }
+
+    ds.add_contributed_values(energy)
+    ds.add_contributed_values(gradient)
+    ds.add_contributed_values(hessian)
+    ds.add_contributed_values(dipole)
+
+    with pytest.raises(KeyError):
+        ds.add_contributed_values(energy)
+
+    ds.save()
+
+    yield client, client.get_collection("dataset", "ds_contributed")
+
+
+def test_dataset_contributed_units(contributed_dataset_fixture):
+    _, ds = contributed_dataset_fixture
+
+    assert "[hartree]" not in ds.get_values(name="Fake Energy").columns[0]
+    assert "[hartree/bohr]" in ds.get_values(name="Fake Gradient").columns[0]
```

Comment:
Interesting, you wouldn't lean towards keeping this in metadata?

Reply:
I'm okay with it going in metadata. Is there a way to attach metadata to columns of a Dataframe? Will it show up in a Jupyter notebook?

Reply:
This is the `self. _column_metadata`. Pandas will be able to hold onto this soonish.